flutter/test/navigation_observer_test.dart

```diff
@@ -0,0 +1,270 @@
+        'to': 'to',
+      });
+    });
+
+    test('route names are null', () {
+      const fromRouteSettings = RouteSettings(name: null, arguments: 'foo');
+
+      const toRouteSettings = RouteSettings(
+        name: null,
+        arguments: {
+          'foo': 123,
+        },
+      );
+
+      // both routes are null
+      final breadcrumb = NavigationBreadcrumb(
+        navigationType: 'didPush',
+        from: fromRouteSettings,
+        to: toRouteSettings,
+      );
+
+      expect(breadcrumb.category, 'navigation');
+      expect(breadcrumb.data, <String, dynamic>{
+        'state': 'didPush',
+        'from_arguments': 'foo',
+        'to_arguments': {
+          'foo': '123',
+        },
+      });
+    });
+  });
+
+  group('SentryNavigationObserver', () {
+    PageRoute route(String name, [Object arguments]) => PageRouteBuilder<void>(
+          pageBuilder: (_, __, ___) => null,
+          settings: RouteSettings(name: name, arguments: arguments),
+        );
+
+    test('Test recording of Breadcrumbs', () {
+      final hub = _FakeHub();
+      final observer = SentryNavigatorObserver(hub: hub);
+
+      final to = route('to', 'foobar');
+      final previous = route('previous', 'foobar');
+
+      observer.didPush(to, previous);
+
+      expect(hub.breadcrumbs.length, 1);
+      expect(hub.breadcrumbs.first.data, <String, dynamic>{
+        'state': 'didPush',
+        'from': 'previous',
+        'from_arguments': 'foobar',
+        'to': 'to',
+        'to_arguments': 'foobar',
+      });
+    });
+
+    test('No arguments', () {
+      final hub = _FakeHub();
+      final observer = SentryNavigatorObserver(hub: hub);
+
+      final to = route('to');
+      final previous = route('previous');
+
+      observer.didPush(to, previous);
+
+      expect(hub.breadcrumbs.length, 1);
+      expect(hub.breadcrumbs.first.data, <String, dynamic>{
+        'state': 'didPush',
+        'from': 'previous',
+        'to': 'to',
+      });
+    });
+
+    test('No arguments & no name', () {
+      final hub = _FakeHub();
+      final observer = SentryNavigatorObserver(hub: hub);
+
+      final to = route(null);
+      final previous = route(null);
+
+      observer.didPush(to, previous);
+
+      expect(hub.breadcrumbs.length, 1);
+      expect(hub.breadcrumbs.first.data, <String, dynamic>{
+        'state': 'didPush',
+      });
+    });
+
+    test('No RouteSettings', () {
+      PageRoute route() => PageRouteBuilder<void>(
+            pageBuilder: (_, __, ___) => null,
+          );
+
+      final hub = _FakeHub();
+      final observer = SentryNavigatorObserver(hub: hub);
+
+      final to = route();
+      final previous = route();
+
+      observer.didPush(to, previous);
+
+      expect(hub.breadcrumbs.length, 1);
+      expect(hub.breadcrumbs.first.data, <String, dynamic>{
+        'state': 'didPush',
+      });
+    });
+  });
+}
+
+/// Used to test if breadcrumbs are really added.
+class _FakeHub implements Hub {
```

Comment:
some people are not fun of `Mocks` and I get that, but if you use the `MockHub` that already exists and `verify` the passed breadcrumb, you could `assert` the values having less boilerplate (all the `UnimplementedError` methods)

Reply:
Done ✅